### PR TITLE
[data] Add TQDM implementation ray.experimental.tqdm_ray; use for Data progress bars

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This adds an experimental `tqdm_ray` module which virtualizes the reporting of tqdm progress state in Ray. It works as follows:
1. The tqdm_ray translates TQDM calls into special json log message events written to worker stdout.
2. The Ray log monitor, instead of copying these log messages directly to the driver stdout, routes these messages to a BarManager singleton.
3. The BarManager manages the positions of progress bars from various Ray tasks / actors, ensuring they don't collide or conflict with each other.

Tl;dr: use `ray.experimental.tqdm_ray` instead of `tqdm` and you won't get pyramid / conflicting tqdm outputs in Ray.

This PR also comes with Datasets integration, which you can enable by setting the RAY_TQDM env var. Try running `RAY_DATASET_USE_STREAMING_EXECUTOR=1 RAY_TQDM=1 python -m ray.experimental.tqdm_ray` for a demo:

Screenshot from 2023-03-09 14-43-18

Limitations:
- Only a subset of tqdm functionality is supported
- Performance may be bad if there are a lot of progress bar updates (we could batch updates in the future)

## Related issues

https://github.com/ray-project/ray/issues/5554
https://github.com/ray-project/ray/issues/9233
https://github.com/skypilot-org/skypilot/issues/262
https://github.com/ray-project/ray/issues/23153
https://github.com/ray-project/ray/issues/23304
https://github.com/ray-project/ray_lightning/issues/176
https://github.com/ray-project/ray/issues/26426
https://github.com/intel-analytics/BigDL/issues/7336

### CODE REVIEW

File: python/ray/data/_internal/progress_bar.py [hunk 1]
```diff
@@ -22,6 +24,60 @@
...
+                        )
+                    else:
+                        bar = self.progress_bars[bar_uuid]
+                        getattr(bar, method)(*args)
+            except Exception as e:
+                print(e)
+
+
+@PublicAPI(stability="beta")
+def enable_centralized_progress_bar() -> None:
```
Comment (ericl) [inline python/ray/data/_internal/progress_bar.py:68]: Another possibility is to record progress in the Ray logs, and using LogMonitor to route/parse progress events for rendering.

File: python/ray/tests/test_tqdm.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    @ray.remote
+    def foo():
+        print("foo started")
+        try:
+            foo = tqdm_ray.tqdm(desc="foo", total=100, position=0)
+            foo.update(42)
+        except Exception as e:
+            print(e)
+        print("foo wait")
+        time.sleep(999)
```
Comment (pcmoritz) [inline python/ray/tests/test_tqdm.py:22]: Was this sleep needed? If yes maybe leave a comment, if no please remove it :)

**Comment by pcmoritz** [comment]
Thanks a lot for doing this! Some local experiments on my mac:
```
from ray.experimental import tqdm_ray

@ray.remote
def foo():
    import time
    foo = tqdm_ray.tqdm(desc="foo", total=100, position=0)
    for i in range(100):
    foo.update(1)
    time.sleep(1.0)

ray.get([foo.remote(), foo.remote(), foo.remote()])
```

Seems to work as advertised :)

**Comment by pcmoritz** [comment]
Small glitch that I see can happen when using ray.get -- but I think it is unrelated to this PR. If we are in an interactive shell we could flush the logs from workers before returning from ray.get.

Screen Shot 2023-03-11 at 2 01 57 PM

**Review by pcmoritz — APPROVED** [review]

**Comment by ericl** [comment]
> Small glitch that I see can happen when using ray.get -- but I think it is unrelated to this PR. If we are in an interactive shell we could flush the logs from workers before returning from ray.get.

Hmm, it might be possible with a shell hook to hide the progress bar, or to add a bit of an exit delay after each shell command. That also gives me an idea on avoiding intermixing of worker output with progress bar output, let me see if we can fix this...

**Review by c21 — APPROVED** [review]
Stamp to unblock.

**Comment by rkooo567** [comment]
cc @xwjiang2010 maybe we can use this for the train output as well?

**Comment by xwjiang2010** [comment]
This looks pretty cool! I want to incorporate it into training/tuning flow. So I did a quick run with the following

```
import ray
from ray.experimental import tqdm_ray

@ray.remote
def foo():
    import time
    foo = tqdm_ray.tqdm(desc="foo", total=100, position=0)
    for i in range(100):
        foo.update(1)
        time.sleep(1.0)

while True:  # simulating driver loop
    try:
        ray.get([foo.remote(), foo.remote(), foo.remote()], timeout=5)
    except ray.exceptions.GetTimeoutError:
        pass
    print("hi~~~~~~~~~")
```

If I have additional messages in driver process to print out other than the progress bar it doesn't look quite right.
How much of work is needed to incorporate that part? 
For AIR output at least, we have a combination of driver log messages in the midst of actor progress bars/actor print messages etc.

**Comment by ericl** [comment]
@xwjiang2010 you can use `ray.experimental.tqdm_ray.safe_print()` which will also avoid corrupting the tqdm bar.

We could potentially change the default system print to safe_print() as well if that would help.

**Comment by juhaszp95** [comment]
Just a quick note to say that when I run the example in the [docs](https://docs.ray.io/en/latest/ray-observability/user-guides/configure-logging.html#distributed-progress-bars-with-tqdm) in a notebook in VS Code, I get no output at all.

**Pull request merged**